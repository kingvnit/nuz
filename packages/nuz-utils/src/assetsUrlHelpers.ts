export const parse = (value: string) => {
  const url = new URL(value)

  const paths = url.pathname.split('/-/')
  if (!paths[1]) {
    throw new Error('URL is invalid')
  }

  const moduleId = paths[0].replace(/^\//, '')
  // @ts-ignore
  const [, version, file] = paths[1].match(/^([^\/]+)\/([\s\S]+)$/i)

  return {
    url,
    paths,
    moduleId,
    version,
    file,
  }
}

export const create = (
  moduleId: string,
  version: string,
  file: string,
  domain: string,
) => `https://${domain}/${moduleId}/-/${version}/${file}`
