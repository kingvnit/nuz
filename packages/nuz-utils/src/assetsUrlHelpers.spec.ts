import * as assetsUrlHelpers from './assetsUrlHelpers'

describe('parse', () => {
  it('Export is function', () => {
    expect(assetsUrlHelpers.parse).toBeInstanceOf(Function)
  })

  describe('Parse valid info', () => {
    const pathnames = [
      [
        '/module/-/0.0.0/other.js',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'other.js',
        },
      ],
      [
        '/module/-/0.0.0/other.js.map',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'other.js.map',
        },
      ],
      [
        '/module/-/0.0.0/styles/other.css',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'styles/other.css',
        },
      ],
      [
        '/module/-/0.0.0/styles/other.css.map',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'styles/other.css.map',
        },
      ],
      [
        '/module/-/0.0.0/images/other.jpg',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'images/other.jpg',
        },
      ],
      [
        '/module/-/0.0.0/images/other',
        {
          moduleId: 'module',
          version: '0.0.0',
          file: 'images/other',
        },
      ],
      [
        '/@scope/module/-/0.0.0/other.js',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'other.js',
        },
      ],
      [
        '/@scope/module/-/0.0.0/other.js.map',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'other.js.map',
        },
      ],
      [
        '/@scope/module/-/0.0.0/styles/other.css',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'styles/other.css',
        },
      ],
      [
        '/@scope/module/-/0.0.0/styles/other.css.map',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'styles/other.css.map',
        },
      ],
      [
        '/@scope/module/-/0.0.0/images/other.jpg',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'images/other.jpg',
        },
      ],
      [
        '/@scope/module/-/0.0.0/images/other',
        {
          moduleId: '@scope/module',
          version: '0.0.0',
          file: 'images/other',
        },
      ],
    ]

    it.each(pathnames)('Pathname `%s`', (pathname, expected) => {
      expect(assetsUrlHelpers.parse(pathname as string)).toMatchObject(expected)
    })
  })

  describe('Fails', () => {
    it('Throw if invalid pathname', () => {
      expect(() => assetsUrlHelpers.parse('/')).toThrowErrorMatchingSnapshot()
    })

    it('Throw if invalid version', () => {
      expect(() =>
        assetsUrlHelpers.parse('/module/-/0-0/images/other.js'),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.parse('/module/-/../images/other.js'),
      ).toThrowErrorMatchingSnapshot()
    })

    it('Throw if invalid module id', () => {
      expect(() =>
        assetsUrlHelpers.parse('/@/-/0.0.0/other.js'),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.parse('/0.0.0/-/0.0.0/other.js'),
      ).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('create', () => {
  it('Export is function', () => {
    expect(assetsUrlHelpers.create).toBeInstanceOf(Function)
  })

  const origin = 'https://static.nuz.app'

  describe('Generate valid', () => {
    const params = [
      [['module', '0.0.0', 'other.js'], `${origin}/module/-/0.0.0/other.js`],
      [
        ['module', '0.0.0', 'other.js.map'],
        `${origin}/module/-/0.0.0/other.js.map`,
      ],
      [
        ['module', '0.0.0', 'styles/other.css'],
        `${origin}/module/-/0.0.0/styles/other.css`,
      ],
      [
        ['module', '0.0.0', 'styles/other.css.map'],
        `${origin}/module/-/0.0.0/styles/other.css.map`,
      ],
      [
        ['module', '0.0.0', 'images/other.jpg'],
        `${origin}/module/-/0.0.0/images/other.jpg`,
      ],
      [
        ['module', '0.0.0', 'images/other'],
        `${origin}/module/-/0.0.0/images/other`,
      ],
      [
        ['@scope/module', '0.0.0', 'other.js'],
        `${origin}/@scope/module/-/0.0.0/other.js`,
      ],
      [
        ['@scope/module', '0.0.0', 'other.js.map'],
        `${origin}/@scope/module/-/0.0.0/other.js.map`,
      ],
      [
        ['@scope/module', '0.0.0', 'styles/other.css'],
        `${origin}/@scope/module/-/0.0.0/styles/other.css`,
      ],
      [
        ['@scope/module', '0.0.0', 'styles/other.css.map'],
        `${origin}/@scope/module/-/0.0.0/styles/other.css.map`,
      ],
      [
        ['@scope/module', '0.0.0', 'images/other.jpg'],
        `${origin}/@scope/module/-/0.0.0/images/other.jpg`,
      ],
      [
        ['@scope/module', '0.0.0', 'images/other'],
        `${origin}/@scope/module/-/0.0.0/images/other`,
      ],
    ]

    it.each(params)('Params %j', (args, expected) => {
      // @ts-ignore
      expect(assetsUrlHelpers.create(...args, origin)).toEqual(expected)
    })
  })

  describe('Fails', () => {
    it('Throw if invalid version', () => {
      expect(() =>
        assetsUrlHelpers.create('module', '0-0', 'filename', origin),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.create('module', '..', 'filename', origin),
      ).toThrowErrorMatchingSnapshot()
    })

    it('Throw if invalid module id', () => {
      expect(() =>
        assetsUrlHelpers.create('@', '0.0.0', 'filename', origin),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.create('0.0.0', '0.0.0', 'filename', origin),
      ).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('createOrigin', () => {
  it('Export is function', () => {
    expect(assetsUrlHelpers.createOrigin).toBeInstanceOf(Function)
  })

  const origin = 'https://static.nuz.app'

  describe('Generate valid', () => {
    const params = [
      [['module', '0.0.0', origin], `${origin}/module/-/0.0.0/`],
      [['module', '0.0.0', origin + '/'], `${origin}/module/-/0.0.0/`],
      [['@scope/module', '0.0.0', origin], `${origin}/@scope/module/-/0.0.0/`],
      [
        ['@scope/module', '0.0.0', origin + '/'],
        `${origin}/@scope/module/-/0.0.0/`,
      ],
    ]

    it.each(params)('Params %j', (args, expected) => {
      // @ts-ignore
      expect(assetsUrlHelpers.createOrigin(...args)).toEqual(expected)
    })
  })

  describe('Fails', () => {
    it('Throw if invalid version', () => {
      expect(() =>
        assetsUrlHelpers.createOrigin('module', '0-0', origin),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.createOrigin('module', '..', origin),
      ).toThrowErrorMatchingSnapshot()
    })

    it('Throw if invalid module id', () => {
      expect(() =>
        assetsUrlHelpers.createOrigin('@', '0.0.0', origin),
      ).toThrowErrorMatchingSnapshot()

      expect(() =>
        assetsUrlHelpers.createOrigin('0.0.0', '0.0.0', origin),
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
