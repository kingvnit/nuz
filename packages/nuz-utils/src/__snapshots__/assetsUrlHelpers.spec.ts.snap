// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create Fails Throw if invalid module id 1`] = `"Module is invalid, value @"`;

exports[`create Fails Throw if invalid module id 2`] = `"Module is invalid, value 0.0.0"`;

exports[`create Fails Throw if invalid version 1`] = `"Version is invalid, value 0-0"`;

exports[`create Fails Throw if invalid version 2`] = `"Version is invalid, value .."`;

exports[`createOrigin Fails Throw if invalid module id 1`] = `"Module is invalid, value @"`;

exports[`createOrigin Fails Throw if invalid module id 2`] = `"Module is invalid, value 0.0.0"`;

exports[`createOrigin Fails Throw if invalid version 1`] = `"Version is invalid, value 0-0"`;

exports[`createOrigin Fails Throw if invalid version 2`] = `"Version is invalid, value .."`;

exports[`parse Fails Throw if invalid module id 1`] = `"Module is invalid, value @"`;

exports[`parse Fails Throw if invalid module id 2`] = `"Module is invalid, value 0.0.0"`;

exports[`parse Fails Throw if invalid pathname 1`] = `"URL is invalid"`;

exports[`parse Fails Throw if invalid version 1`] = `"Version is invalid, value 0-0"`;

exports[`parse Fails Throw if invalid version 2`] = `"Version is invalid, value .."`;
